Return value from protect()

berdroid opened the pull request:
The Mutex.protect() method is modified to return a Future that completes with the value returned by the criticalSection function. This adds useful functionality for many use cases.
Also added analyzer exception for dead code in 'lock released on exception' test.

Test cases have been added, and all existing tests remain green.

berdroid committed:
Return value from protect()

The Mutex.protect() method is modified to return a Future that completes with the value returned by the criticalSection function. This adds useful functionality for many use cases.
Also added analyzer exception for dead code in 'lock released on exception' test

berdroid commented:
BTW Thanks for this small yet very useful package. Well done!

hoylen commented:
Thank you very much for that really cool feature.

And it doesn't seem to break any existing code too. So I'll push out a new release to pub.dev with it too.

hoylen commented:
Actually, this feature has opened up an issue.

You've defined the _criticalSection_ as a function returning `T`. But it really should be a function that returns a `Future<T>`. However, defining it to return a `Future<T>` breaks the current tests, since they use synchronous functions for testing. I think in all/most real-world situations the critical section will be an asynchronous function (otherwise, it wouldn't be critical). So I think changing it to `Future<T>` should not affect many people, but it is a breaking change so it should be published with a new major version number.

The alternative is to define it as a function that returns `FutureOr<T>`, but that seems less correct. I think there is value in correctly defining it as a `Future<T>` so people are alerted to the fact that their critical section is not critical when it only contains synchronous code.

berdroid commented:
You are right, I should have considered that more carefully. Actually the code I initially write in my project at work did just that. I have updated my clone of the repo to do this correctly.  In order to resolve the issue with existing tests I modified those to use async functions for the critical section.  I also added some new tests to ensure that all types including void and nullable types can be returned.

I'll open a new pull request for that.
